### DISCUSSION TITLE
Quasar component wrapper

### DISCUSSION THREAD

**Discussion by ArjanAkkermans** [quasarframework/quasar#16143] [General - Components / Directives / etc]
Hi,

I am somewhat new to vue/quasar and was wondering how one should go about implementing a component wrapper. I realize the question is not only related to quasar, but  I wanted to ask experience here about whether others are using wrapper components around Quasar components. For now the main issue I am facing is how to handle around the fact that unspecified boolean input arguments are changed from undefined to false (https://github.com/vuejs/vue/issues/4792). With my current setup, and I am unable to find a satisfactory solution for this. 

General answers/experience are also welcome, do you use wrappers for Quasar components, and have there been any issues that you encountered? Can my custom wrapper be improved / am I missing something else?

To have a concrete example: I want to implement a wrapper for the QSelect component which enables filtering of the options, such that this logic is encapsulated into a single place. The following custom component achieves this. 

````
<script setup lang="ts">
  import { QSelectProps, QSelect } from "quasar";
  import { computed, ref } from "vue";

  const props = withDefaults(defineProps<QSelectProps>(), {});

  const filteredOptions = ref(props.options);

  type UpdateFunctionType = (fn: () => void) => void;
  function updateFilteredOptions(val: string, update: UpdateFunctionType) {
    if (val === "") {
      update(() => {
        filteredOptions.value = props.options;
      });
      return;
    }
    update(() => {
      const needle = val.toLocaleLowerCase();
      filteredOptions.value =
        props.options?.filter(
          (filteredOption) => filteredOption.label.toLowerCase().indexOf(needle) > -1
        ) ?? [];
    });
  }

  const qSelectProps = computed(() => {
    return {
      ...props,
      options: filteredOptions.value,
      useInput: true,
    };
  });
</script>

<template>
  <QSelect ref="select" v-bind="qSelectProps" @filter="updateFilteredOptions"> </QSelect>
</template>
````

In this case, calling this custom component with a number of QSelectProps specified is not similar to calling the QSelect component directly, and that is caused by the 'withDefaults(defineProps<QSelectProps>(), {});' function setting the passed booleans to false. E.G. the 'dark' property behaves different when it is set to false or undefined, and by using this wrapper the value is unintentionally changed. To solve this issue it would suffice to specify the default argument, e.g. 'withDefaults(defineProps<QSelectProps>(), {dark: undefined});', but ideally I shouldn't need to do this.

There are two possible workarounds for this issue as a whole, firstly I could simply remove the input property type QSelectProps, but then no hints/autocompletion are given when using the custom component. Another solution would be to specify that all the boolean values of QSelect should be undefined, however the (small) issue with that approach is that it both bloats the code, and requires changes might new boolean input arguments be introduced by the quasar library.

**Comment by eduardoconviso** [comment]
I am having the same issue. Any news?

**Reply by ArjanAkkermans** [reply]
Unfortunately I have not yet found a satisfactory answer but I can share what my team decided on ( the most experienced person has around one year of vue experience so definitely I am aware we might not know everything ;-))

We still use the wrapper of a QSelect with input props of type QSelectProps. To get around the issue of boolean attributes being changed from undefined to false when passing as arguments we construct a defaults object based on the defaults of QSelect which is retreived using the following piece of code:

```
export function defaultPropsExtractor(props: any, ignoreProps: string[] = []) {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const defaultProps: Record<string, unknown> = {};
  for (const key in props) {
    if ("default" in props[key] && !ignoreProps.includes(key)) {
      defaultProps[key] = props[key]["default"];
    }
  }
  return defaultProps;
}
```

Where ignoreProps is used to specify props that should have a default value (e.g. props overridden in the wrapper). 

Specifically this function is called from  a separate script, so it is loaded on module level (non-setup) because the default prop values need to be a `static' data before the setup script is executed, the code calling it looks as following: (note the two seperate script blocks!)

```
<script lang="ts">
  import { defaultPropsExtractor } from "@/utils/DefaultPropsExtractor";
  const defaultPropsQSelect = defaultPropsExtractor(QSelect.props, [
    "useInput", // make sure to not include useInput default, as it is omitted from the defineProps
  ]);
</script>

<script setup lang="ts">
---------------------
</script>

<template>
------
<template>
```



An additional thing to point out specifically for my use case, using a filter on the QSelect, is that a QSelect with filter is not really identical as a QSelect without filter, as the filter can take up more space and therefore you should not expect to be able to replace all your QSelects with a QSelect with filter without any regressions!


I would be happy to be convinced otherwise, but for now my conclusion would be not to use custom component wrappers as the hassle is not worth the benefits, for my specific use case I would rather have implemented the filtering logic at some specific places (probably using some kind of composable to handle part of the logic) as opposed to generically implementing the QSelect with filter.
